sycl/include/sycl/ext/oneapi/sub_group_mask.hpp

```diff
@@ -94,11 +94,15 @@ struct sub_group_mask {
   bool none() const { return count() == 0; }
   uint32_t count() const {
     unsigned int count = 0;
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
+    asm("popc.b32 %0, %1;" : "=r"(count) : "r"(Bits));
+#else
     auto word = (Bits & valuable_bits(bits_num));
     while (word) {
       word &= (word - 1);
       count++;
     }
+#endif
```

Comment:
Yes this would be good. I suppose this corresponds to `__spirv_GroupNonUniformBallotBitCount(
      __spv::Scope::Subgroup, (int)__spv::GroupOperation::Reduce,
      OCLMask);` ?

I think I may as well generalize the `GroupNonUniformBallotBitCount` cuda impl now. then call it in the above code. Then when spirv impl is ready we won't need any macro control flow in the above code.

Reply:
Yes, I think that's right.

It does mean having to construct a 128-bit mask from a 32-bit mask just to call the intrinsic, but hopefully the compiler can optimize the unused components away.

Reply:
OK I just want to be sure I can do it in a backend agnostic way. since amd is using sometimes 64 bits will it be OK for general backends if I `std::memcpy` `Bits` to the 128-bit mask? i.e will this result in the expected 128-bit mask format in the `GroupNonUniformBallotBitCount` for all backends?

Reply:
I guess there not a spirv function corresponding to `IdToMaskPosition`?

Reply:
> I guess there not a spirv function corresponding to `IdToMaskPosition`?

I don't think so.  I looked for one and couldn't find it, which is why I wrote that naive loop.